docs/protocol/specifications/validation.md

```diff
@@ -294,55 +131,46 @@ Note: requirement for the input and output sums to be below 2<sup>63</sup> impli
+    1. [Apply the transaction](#apply-transaction) using the input block’s header to blockchain state `S`, yielding a new state `S′`.
+    2. If transaction failed to be applied (did not change blockchain state), halt and return the input blockchain state unchanged.
+    3. Replace `S` with `S′`.
+4. Test that [assets merkle root](blockchain.md#assets-merkle-root) of `S` is equal to the assets merkle root declared in the block header; if not, halt and return blockchain state unchanged.
+5. Remove elements of the nonce set in `S` where the expiration timestamp is less than the block’s timestamp, yielding a new state `S′`.
+6. Return the state `S’`.
 
-### Apply Transaction
+
+### Apply transaction
 
 **Inputs:**
 
 1. transaction,
-2. blockchain state.
+2. block header,
+3. blockchain state.
 
 **Output:** blockchain state.
 
 **Algorithm:**
 
-1. For each spend input with asset version 1 in the transaction:
-    1. Delete the previous [output ID](data.md#output-id) from S, yielding a new state S′.
-    2. Replace S with S′.
-2. For each output with asset version 1 in the transaction:
-    1. Add that output’s [output ID](data.md#output-id) to S, yielding a new state S′.
-    2. Replace S with S′.
-3. For all [asset version 1 issuance inputs](data.md#asset-version-1-issuance-commitment) with non-empty *nonce* string:
-    1. Compute the [issuance hash](data.md#issuance-hash) H.
-    2. Add (H, transaction maximum timestamp) to the issuance memory in S, yielding a new state S′.
-    3. Replace S with S′.
-4. Return S.
-
-
-### Evaluate predicate
-
-**Inputs:**
-
-1. VM version,
-2. program,
-3. list of program arguments.
-
-**Output:** true or false.
-
-**Algorithm:**
-
-1. If the [VM version](vm1.md#versioning) is > 1, halt and return true.
-2. [Create a VM with initial state](vm1.md#vm-state).
-3. [Prepare VM](vm1.md#prepare-vm).
-4. Set the VM’s program to the predicate program and execute [Verify Predicate](vm1.md#verify-predicate) operation. If it fails, halt and return false.
-5. Return true.
+1. Validate transaction using the checks below. If any check fails, halt and return the input blockchain state unchanged.
+    1. If the block header version is 1, verify that transaction version is equal to 1.
+    2. If the `Mintime` is greater than zero: verify that it is less than or equal to the block header timestamp.
+    3. If the `Maxtime` is greater than zero: verify that it is greater than or equal to the block header timestamp.
+    4. [Validate transaction header](blockchain.md#transaction-header-validation).
+2. Let `S` be the input blockchain state.
+3. For each visited [nonce entry](blockchain.md#nonce) in the transaction:
+    1. If [nonce ID](blockchain.md#entry-id) is already stored in the nonce set of the blockchain state, halt and return the input blockchain state unchanged.
+    2. Add ([nonce ID](blockchain.md#entry-id), transaction maxtime) to the nonce set in `S`, yielding a new state `S′`.
```

Comment:
Is the transaction maxtime the right expiration time for the nonce, or should it be the nonce's own timerange.maxtime? (Or the max of those two numbers? Or something else?)